From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Sat, 17 Jan 2026 11:45:51 +0000
Subject: [PATCH] Use original gameprofile for dupe profile checks

The server checks for the player profiles id when having a new
player join the server. While in proxy online mode, it also checks based
on name. This causes players nicked via the player profile API to be
kicked.

Instead, check the initial game profile of the player during join.

diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 53522a015a49e46533b5862f251d0ea1ffcce7db..e22d74ab9388941cbd2787ef28b6b5c004c95c47 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -576,7 +576,7 @@ public abstract class PlayerList {
         Set<ServerPlayer> dupes = Sets.newIdentityHashSet();
 
         for (ServerPlayer player : this.players) {
-            if (player.getUUID().equals(playerId) || (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode() && player.getGameProfile().name().equalsIgnoreCase(profile.name()))) { // Paper - validate usernames
+            if (player.getUUID().equals(playerId) || (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode() && player.originalGameProfile.name().equalsIgnoreCase(profile.name()))) { // Paper - validate usernames // KTP - use original game profile for comparison
                 dupes.add(player);
             }
         }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 5d4b3ed09fe10f545d0720b2919f0d6c822d5196..cc658403cdcb2ba52ee2dd1ff5eb5bc7164da079 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -184,6 +184,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         return (org.bukkit.craftbukkit.entity.CraftHumanEntity) super.getBukkitEntity();
     }
     // CraftBukkit end
+    public final GameProfile originalGameProfile; // KTP - original player profile storage
 
     public Player(final Level level, final GameProfile gameProfile) {
         super(EntityTypes.PLAYER, level);
@@ -192,6 +193,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         this.inventory = new Inventory(this, this.equipment);
         this.inventoryMenu = new InventoryMenu(this.inventory, !level.isClientSide(), this);
         this.containerMenu = this.inventoryMenu;
+        this.originalGameProfile = gameProfile; // KTP - original player profile storage
     }
 
     @Override
